### PR TITLE
Linux bugfixes

### PR DESCRIPTION
**Pull request by aarzilli**
I see you already covered my d95f9ca, 
c2a346e is just an optimization,
b7e6e27 is a workaround for weird behaviour of wait on linux

### CODE REVIEW

File: proc/proc_linux.go [hunk 1]
```diff
@@ -270,6 +272,10 @@ func (dbp *Process) trapWait(pid int) (*Thread, error) {
 			}
 			continue
 		}
+		if th == nil {
+			// Sometimes we get an unknown thread, ignore it?
```
Comment (derekparker) [inline proc/proc_linux.go]: This doesn't seem right. The only time we should get an unknown thread is when a new one is cloned and it's the first time we're hearing about it. Was this crashing?

File: proc/proc_linux.go [hunk 2]
```diff
@@ -304,14 +310,45 @@ func stopped(pid int) bool {
...
+
+		for {
+			wpid, err := sys.Wait4(pid, &s, sys.WNOHANG|sys.WALL|options, nil)
+			if err != nil {
+				return 0, nil, err
+			}
+			if wpid != 0 {
+				return wpid, &s, err
+			}
+
```
Comment (derekparker) [inline proc/proc_linux.go]: Kill whitespace.

File: proc/breakpoints.go [hunk 3]
```diff
@@ -139,6 +139,11 @@ func (dbp *Process) setBreakpoint(tid int, addr uint64, temp bool) (*Breakpoint,
 	return dbp.Breakpoints[addr], nil
 }
 
+func (dbp *Process) writeSoftwareBreakpoint(thread *Thread, addr uint64) error {
```
Comment (derekparker) [inline proc/breakpoints.go]: I like this, but rename to setSoftwareBreakpoint to be more consistent with setHardwareBreakpoint.

**Comment by derekparker** [comment]
This is great, thank you. Just some minor feedback.

File: proc/proc_linux.go (hunk 1 above)
Comment (aarzilli) [inline proc/proc_linux.go]: yes it was crashing, a nil pointer dereference at th.running = false later on, yes this is weird

File: proc/breakpoints.go (hunk 3 above)
Comment (aarzilli) [inline proc/breakpoints.go]: I will do that if you insist but they are not symmetric, setHardwareBreakpoint is a complete operation, to set a software breakpoint we have to read the original instructions and then write the breakpoint and writeSoftwareBreakpoint only does the second part.

Comment (derekparker) [inline proc/breakpoints.go]: Good point, disregard this comment.

**Comment by derekparker** [comment]
I rebased this for you since it was a bit tricky due to my code changes and merged it in. Thanks for the patch.

**Pull request closed without merging**